### PR TITLE
Introduce baseline JIT compiler

### PR DESCRIPTION
**Pull request by qwe661234**
The baseline JIT compiler design involves tracing an EBB when its usage frequency exceeds a predetermined threshold. We then utilize a code generator to convert the instruction sequence into C code. The C code is subsequently compiled using mir, and the resulting target machine code is stored in the code cache for future utilization.

The primary objective of introducing the baseline JIT compiler is to enhance the execution speed of RISC-V instructions with minimal modification. This implementation requires three additional components: a code generator, mir compiler, and code cache, while maintaining the original design of the interpreter. Furthermore, this baseline JIT compiler serves as the foundational target for future improvements.

|Metric|  master| baseline JIT|Speedup|
|--------| -------- | -------- | -------- | 
| CoreMark | 1352.843 (Iterations/Sec)|1740.881 (Iterations/Sec) |+28.6％|
| dhrystone  |1146 DMIPS |2368.33DMIPS |+106%|

### CODE REVIEW

File: src/cache.c [hunk 1]
```diff
@@ -16,27 +17,31 @@
...
 static uint32_t cache_size, cache_size_bits;
 static struct mpool *cache_mp;
 
 #if RV32_HAS(ARC)
-/* The Adaptive Replacement Cache (ARC) improves the traditional LRU strategy
- * by dividing the cache into two lists: T1 and T2. T1 follows the LRU
- * strategy, while T2 follows the LFU strategy. Additionally, ARC maintains two
- * ghost lists, B1 and B2, which store replaced entries from the LRU and LFU
- * lists, respectively.
+/*
```
Comment (jserv) [inline src/cache.c]: Minimize the necessary changes. Don't change this comment.

**Review by jserv — CHANGES_REQUESTED** [review]
Resolve the issues raised by CodeQL in advance.

File: src/cache.c [hunk 2]
```diff
@@ -16,27 +17,31 @@
 #define GOLDEN_RATIO_32 0x61C88647
 #define HASH(val) \
     (((val) * (GOLDEN_RATIO_32)) >> (32 - (cache_size_bits))) & (cache_size - 1)
-
 /* THRESHOLD is set to identify hot spots. Once the frequency of use for a block
- * exceeds the THRESHOLD, the JIT compiler flow is triggered.
- */
-#define THRESHOLD 1000
+ * exceeds the THRESHOLD, the JIT compiler flow is triggered. */
```
Comment (jserv) [inline src/cache.c]: Minimize the necessary changes.

File: src/codegen.c [hunk 3]
```diff
@@ -0,0 +1,1298 @@
...
+        RISCV_INSN_LIST
+#undef _
+        return true;
+    }
+    return false;
+}
+typedef void (*gen_func_t)(riscv_t *, const rv_insn_t *, char *, uint32_t);
+static gen_func_t dispatch_table[rv_insn_qty];
+static char funcbuf[128] = {0};
+#define RVOP(inst, code)                                                  \
```
Comment (jserv) [inline src/codegen.c]: Instead of duplicating `RVOP` and instruction executions, try to provide JIT-aware `RVOP` macro which concatenates the string for code generation.

**Review by jserv — CHANGES_REQUESTED** [review]
Check the `jit_template.inc` in branch `wip/jit` and eliminate the code duplication between `execute.c` and `codegen.c`.

**Comment by jserv** [comment]
Resolve #136 before integrating baseline JIT compiler.

File: Makefile [hunk 4]
```diff
@@ -78,6 +78,21 @@ gdbstub-test: $(BIN)
 	$(Q).ci/gdbstub-test.sh && $(call notice, [OK])
 endif
 
+ENABLE_JIT ?= 1
```
Comment (jserv) [inline Makefile]: Default to `0` (disabled).

File: Makefile [hunk 5]
```diff
@@ -78,6 +78,21 @@ gdbstub-test: $(BIN)
...
+ifeq ($(call has, JIT), 1)
+CFLAGS += -I./mir
+LDFLAGS += mir/libmir.a -lpthread
+
+OBJS_EXT += codegen.o compile.o
+
+mir/GNUmakefile:
+	git submodule update --init $(dir $@)
+	
+mir/libmir.a: mir/GNUmakefile
```
Comment (jserv) [inline Makefile]: For the consistency of source code layout, place MIR directory under `src`. That is, specify `src/mir` here.

File: src/riscv_private.h [hunk 6]
```diff
@@ -119,8 +126,12 @@ struct riscv_internal {
     uint32_t csr_mip;
     uint32_t csr_mbadaddr;
 
-    bool compressed;       /**< current instruction is compressed or not */
+    bool compressed; /**< current instruction is compressed or not */
+#if !RV32_HAS(JIT)
     block_map_t block_map; /**< basic block map */
+#else
+    struct cache *cache;
```
Comment (jserv) [inline src/riscv_private.h:133]: To minimize the code changes, is it possible to stick to `block_map_t` rather than `struct cache`?

Comment (qwe661234) [reply src/riscv_private.h:133]: It is difficult to stick to `block_map_t`, because the structure of ARC cache is totally different from `block_map_t`.

File: .vscode/settings.json [hunk 7]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (jserv) [inline .vscode/settings.json]: Don't check in anything related to `.vscode`.

File: Makefile [hunk 8]
```diff
@@ -109,9 +124,15 @@ $(OUT)/%.o: src/%.c
 	$(VECHO) "  CC\t$@\n"
 	$(Q)$(CC) -o $@ $(CFLAGS) -c -MMD -MF $@.d $<
 
+ifeq ($(call has, JIT), 0)
 $(BIN): $(OBJS)
 	$(VECHO) "  LD\t$@\n"
 	$(Q)$(CC) -o $@ $^ $(LDFLAGS)
+else
+$(BIN): src/mir/libmir.a $(OBJS)
```
Comment (jserv) [inline Makefile]: You can simply append `src/mir/libmir.a` to `$(OBJS)`.

File: src/codegen.c [hunk 9]
```diff
@@ -0,0 +1,1238 @@
...
+RVOP(auipc, {
+    sprintf(funcbuf, "  rv->X[%u] = %u + rv->PC;\n", ir->rd, ir->imm);
+    strcat(gencode, funcbuf);
+})
+
+RVOP(jal, {
+    strcat(gencode, "  pc = rv->PC;\n");
+    UPDATE_PC(ir->imm);
+    if (ir->rd) {
+        sprintf(funcbuf, "  rv->X[%u] = pc + %u;\n", ir->rd, ir->insn_len);
```
Comment (jserv) [inline src/codegen.c]: Don't invoke `sprintf` in each implementation of RISC-V instructions. Instead, wrap them in proper macros or inline functions.

File: src/elfdef.h [hunk 10]
```diff
@@ -0,0 +1,78 @@
+#include <stdint.h>
```
Comment (jserv) [inline src/elfdef.h]: Don't create new file `elfdef.h`. Instead, consolidate the existing `elf.[ch]`.

**Comment by jserv** [comment]
The proposed changes aim to achieve the following objectives:
1. Elimination of code duplication between `emulate.c` and `codegen.c`.
2. Reuse of the RV32 interpreter implementation for the baseline JIT.

These refinements focus on improving code organization and minimizing duplication. By reusing the existing RV32 interpreter implementation, we can expedite the development of the JIT compiler while maintaining consistency between the interpreter and JIT modes of execution.

The new branch `wip/jit` lays the groundwork for the future development of a baseline JIT compiler. By decoupling the RV32 instructions, we create reusable implementation templates that can be leveraged for code generation in the JIT compiler. This decoupling enables straightforward and flexible code generation strategies, eliminating the need for code duplication between the interpreter and MIR-based JIT compiler.

Assuming the existence of a dedicated shell script capable of generating C-string operations for each `RVOP` entry in `src/codegen.c` from `src/rv32_template.c`, the following procedure can be followed:
1. Enable the JIT compiler in the top-level configurations.
2. Generate `src/codegen.c` from `src/rv32_template.c` using the aforementioned script.
3. The file `src/emulate.c` includes the generated `src/codegen.c` for the JIT execution path.
4. The runtime determines the execution path, whether through the interpreter or JIT compilation, based on heuristic algorithms or thresholds.

This pull request should be on top of branch `wip/jit`, and you will have to rebase it and rework.

File: src/cache.c [hunk 11]
```diff
@@ -286,6 +305,12 @@ cache_t *cache_create(int size_bits)
         mpool_create(cache_size * sizeof(lfu_entry_t), sizeof(lfu_entry_t));
 #endif
     cache->capacity = cache_size;
+#if RV32_HAS(JIT)
+    cache->jitcode = (uint8_t *) mmap(NULL, CODE_CACHE_SIZE,
```
Comment (jserv) [inline src/cache.c]: The casting from `void *` to `uint8_t *` is not necessary in C programming language.

**Review by jserv — CHANGES_REQUESTED** [review]
Rebase latest `wip/jit` branch rather than `master` branch.

**Review by jserv — CHANGES_REQUESTED** [review]
Rebase latest `wip/jit` branch rather than `master` branch.

File: Makefile [hunk 12]
```diff
@@ -78,6 +78,26 @@ gdbstub-test: $(BIN)
...
 endif
 
+ENABLE_JIT ?= 0
+$(call set-feature, JIT)
+ifeq ($(call has, JIT), 1)
+CFLAGS += -I./src/mir
+LDFLAGS += src/mir/libmir.a -lpthread
+OBJS_EXT += codegen.o compile.o
+src/rv32_jit_template.c:
+	$(Q)python ./tools/jit_template_grnerator.py
```
Comment (jserv) [inline Makefile]: Rename to `tools/gen-jit-template.py`

File: src/riscv.h [hunk 13]
```diff
@@ -7,6 +7,22 @@
 #include <stdbool.h>
 #include <stdint.h>
 
+#if RV32_HAS(EXT_F)
```
Comment (jserv) [inline src/riscv.h]: Don't do that. The file `riscv.h` is supposed to be a public header, and there should be no implementation details.

File: .gitmodules [hunk 14]
```diff
@@ -4,3 +4,6 @@
 [submodule "mini-gdbstub"]
 	path = src/mini-gdbstub
 	url = https://github.com/RinHizakura/mini-gdbstub
+[submodule "mir"]
+	path = mir
```
Comment (jserv) [inline .gitmodules]: Should be `src/mir`.

File: .gitmodules [hunk 15]
```diff
@@ -4,3 +4,6 @@
 [submodule "mini-gdbstub"]
 	path = src/mini-gdbstub
 	url = https://github.com/RinHizakura/mini-gdbstub
+[submodule "src/mir"]
+	path = src/mir
+	url = https://github.com/vnmakarov/mir.git
```
Comment (jserv) [inline .gitmodules]: Trim the trailing `.git` since it is not necessary.

File: Makefile [hunk 16]
```diff
@@ -78,6 +78,30 @@ gdbstub-test: $(BIN)
...
+	$(MAKE) --quiet -C src/mir
+
+$(OUT)/codegen.o: src/codegen.c src/rv32_jit_template.c
+	$(VECHO) "  CC\t$@\n"
+	$(Q)$(CC) -o $@ $(CFLAGS) -c -MMD -MF $@.d $<
+
+$(OUT)/compile.o: src/compile.c src/mir/libmir.a
+	$(VECHO) "  CC\t$@\n"
+	$(Q)$(CC) -o $@ $(CFLAGS) -c -MMD -MF $@.d $<
+
```
Comment (jserv) [inline Makefile]: Remove this blank line.

File: Makefile [hunk 17]
```diff
@@ -78,6 +78,30 @@ gdbstub-test: $(BIN)
 	$(Q).ci/gdbstub-test.sh && $(call notice, [OK])
 endif
 
+ENABLE_JIT ?= 0
+$(call set-feature, JIT)
+ifeq ($(call has, JIT), 1)
+CFLAGS += -I./src/mir
+LDFLAGS += src/mir/libmir.a -lpthread
+OBJS_EXT += codegen.o compile.o
+src/rv32_jit_template.c:
```
Comment (jserv) [inline Makefile]: Simply call it `jit_template.c` for short.

File: src/cache.c [hunk 18]
```diff
@@ -226,6 +244,7 @@ static inline void hlist_del_init(struct hlist_node *n)
          pos = hlist_entry_safe((pos)->member.next, type, member))
 #endif
 
+
```
Comment (jserv) [inline src/cache.c]: Don't do this. Minimize the necessary changes.

File: src/codegen.c [hunk 19]
```diff
@@ -0,0 +1,385 @@
+/*
```
Comment (jserv) [inline src/codegen.c]: Can you move the handwritten part of `src/codegen.c` into `src/compile.c`? Therefore, `src/codegen.c` would be completely generated by tools, implying that we can ignore it in git tree.

File: src/compile.c [hunk 20]
```diff
@@ -0,0 +1,129 @@
+#include <stdio.h>
```
Comment (jserv) [inline src/compile.c:6]: Add copyright notice.

File: src/compile.c [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
```
Comment (jserv) [inline src/compile.c]: Header `<sys/mman.h>` is not needed here.

File: src/compile.c [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+{
+    jit_item_t *item = data;
+    return item->curr >= item->code_size ? EOF : item->code[item->curr++];
+}
+
+#define DLIST_ITEM_FOREACH(modules, item)                     \
+    for (item = DLIST_HEAD(MIR_item_t, modules->items); item; \
+         item = DLIST_NEXT(MIR_item_t, item))
+
+void *import_resolver(const char *name)
```
Comment (jserv) [inline src/compile.c]: Mark it as `static`.

File: src/compile.c [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+        if (!strcmp(name, func_list[i].name))
+            return func_list[i].func;
+    }
+    return NULL;
+}
+
+static riscv_jit_t *jit = NULL;
+static jit_item_t *jit_ptr = NULL;
+
+uint8_t *compile(riscv_t *rv)
```
Comment (jserv) [inline src/compile.c]: Is it an internal function? If so, mark it as `static`.

File: src/compile.h [hunk 24]
```diff
@@ -0,0 +1,12 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+#pragma once
+#include <stdint.h>
+
+#include "riscv.h"
+#include "riscv_private.h"
```
Comment (jserv) [inline src/compile.h]: Is header `riscv_private.h` needed?

File: src/decode.h [hunk 25]
```diff
@@ -171,6 +171,7 @@ enum {
 #define _(inst, can_branch) rv_insn_##inst,
     RISCV_INSN_LIST
 #undef _
+        rv_insn_qty,
```
Comment (jserv) [inline src/decode.h]: Rename to `N_RV_INSN` for consistency.

File: src/riscv.c [hunk 26]
```diff
@@ -148,3 +167,135 @@ void rv_reset(riscv_t *rv, riscv_word_t pc)
 
     rv->halt = false;
 }
+
+#if defined(__APPLE__)
+inline int isinff(float x)
```
Comment (jserv) [inline src/riscv.c]: Mark it as `static inline`.

File: src/riscv.c [hunk 27]
```diff
@@ -148,3 +167,135 @@ void rv_reset(riscv_t *rv, riscv_word_t pc)
...
+{
+    return __builtin_fabsf(x) == __builtin_inff();
+}
+inline int isnanf(float x)
```
Comment (jserv) [inline src/riscv.c]: Ditto. Mark it as `static inline`.

File: src/riscv.h [hunk 28]
```diff
@@ -7,6 +7,16 @@
...
```
Comment (jserv) [inline src/riscv.h]: Move these into `riscv_private.h`. Always ensure the simplicity of public headers.

File: src/riscv.h [hunk 29]
```diff
@@ -167,6 +177,26 @@ bool rv_has_halted(riscv_t *rv);
 /* return the flag of outputting exit code */
 bool rv_enables_to_output_exit_code(riscv_t *rv);
 
+#if RV32_HAS(Zicsr)
```
Comment (jserv) [inline src/riscv.h]: Ditto. Don't touch public headers.

File: tools/gen-jit-template.py [hunk 30]
```diff
@@ -0,0 +1,58 @@
+import operator
```
Comment (jserv) [inline tools/gen-jit-template.py]: Enforce [PEP 8 – Style Guide for Python Code](https://peps.python.org/pep-0008/).

File: src/compile.c [hunk 31]
```diff
@@ -0,0 +1,507 @@
...
+    void *func;
+} func_obj_t;
+
+#define SET_SIZE 1024
+
+typedef struct {
+    uint32_t table[SET_SIZE][32];
+} set_t;
+
+static inline uint32_t hash(uint32_t key)
```
Comment (jserv) [inline src/compile.c]: Why did you select this hash function?

File: src/compile.c [hunk 32]
```diff
@@ -0,0 +1,507 @@
...
+{
+    return (key >> 1) & (SET_SIZE - 1);
+}
+
+void set_reset(set_t *set)
```
Comment (jserv) [inline src/compile.c]: Does it make sense to expose set related symbols?

File: src/compile.c [hunk 33]
```diff
@@ -0,0 +1,507 @@
...
+            break;
+        ir++;
+    }
+    if (ir->branch_untaken && !set_has(set, ir->branch_untaken->pc))
+        trace_ebb(rv, gencode, ir->branch_untaken, set);
+    if (ir->branch_taken && !set_has(set, ir->branch_taken->pc))
+        trace_ebb(rv, gencode, ir->branch_taken, set);
+}
+
+#define PROLOGUE                                                          \
```
Comment (jserv) [inline src/compile.c]: Is it possible to generate `PROLOGUE` from C source file?

File: tools/gen-jit-template.py [hunk 34]
```diff
@@ -0,0 +1,156 @@
...
+impl = []
+for i in range(len(str2)):
+    # remove comment
+    tmp = re.sub(r'/\*[\s|\S]+?\*/\n', "", str2[i])
+    op.append(tmp[5:tmp.find(',')])
+    impl.append(tmp[tmp.find('{') + 1:-2])
+
+f.close()
+# generate jit template
+f = open("src/jit_template.c", 'w')
```
Comment (jserv) [inline tools/gen-jit-template.py]: Output to stdout by default.

File: tools/gen-jit-template.py [hunk 35]
```diff
@@ -0,0 +1,156 @@
...
+    "clw",
+    "csw",
+    "cjal",
+    "cj",
+    "cbeqz",
+    "cbnez",
+    "fuse3",
+    "fuse4"]
+# check enabled extension in Makefile
+f = open('Makefile', 'r')
```
Comment (jserv) [inline tools/gen-jit-template.py]: Don't do that. Instead, the build rules should generate something corresponded to the current configuration. Then, tools can take the generated configuration file(s).

File: src/cache.c [hunk 36]
```diff
@@ -541,3 +568,117 @@ void cache_free(cache_t *cache, void (*callback)(void *))
...
+        if (entry->key == key)
+            return cache->jitcode + entry->offset;
+    }
+#endif
+    return NULL;
+}
+
+static inline uint64_t align_to(uint64_t val, uint64_t align)
+{
+    if (!align)
```
Comment (jserv) [inline src/cache.c]: Add `unlikely` here.

File: src/emulate.c [hunk 37]
```diff
@@ -615,7 +504,27 @@ void rv_step(riscv_t *rv, int32_t cycles)
...
+        if (block->hot)
+            code = code_cache_lookup(rv->cache, block->pc_start);
+        if (!code) {
+            /* check if using frequency of block exceed threshold */
+            if ((block->hot = cache_hot(rv->cache, block->pc_start)))
+                code = block_compile(rv);
+        }
+        if (block->hot) {
+            /* execute machine code */
+            /* clang-format off */
```
Comment (jserv) [inline src/emulate.c]: Why did you add clang-format tweak here?

File: src/riscv_private.h [hunk 38]
```diff
@@ -137,3 +148,33 @@ static inline uint32_t sign_extend_b(const uint32_t x)
 {
     return (int32_t) ((int8_t) x);
 }
+
+#if RV32_HAS(EXT_F)
+#include <math.h>
+#include "softfloat.h"
+
+#if defined(__APPLE__)
```
Comment (jserv) [inline src/riscv_private.h:156]: This part is no longer needed.

File: src/riscv.c [hunk 39]
```diff
@@ -148,3 +167,135 @@ void rv_reset(riscv_t *rv, riscv_word_t pc)
 
     rv->halt = false;
 }
+
+#if defined(__APPLE__)
```
Comment (jserv) [inline src/riscv.c:171]: Replace with the following:
```c
if RV32_HAS(EXT_F) && defined(__APPLE__)

```

Comment (jserv) [inline src/riscv.c]: Remove the inline functions `isinff` and `isnanf`. The function definitions should appear in `src/riscv_private.h`.

**Comment by jserv** [comment]
TODO: benchmarking on the following machine:
* eMag 8180 Arm server w/ GNU/Linux
* Apple M1 (or other Apple Silicon) w/ macOS

We have to ensure the benefits.

File: src/emulate.c (hunk 37 above)
Comment (qwe661234) [reply src/emulate.c]: After using `clang-format-12`, this line would become `(exec_block_func_t) code) (rv, block->ir)`. However, it fails the coding style check because the expected format is without a space `(exec_block_func_t) code)(rv, block->ir)`.

File: src/riscv_private.h (hunk 38 above)
Comment (qwe661234) [reply src/riscv_private.h:156]: Except for Apple, the libraries of other computer architectures has provided the `isinff` and `isnanf` functions

File: src/emulate.c (hunk 37 above)
Comment (jserv) [reply src/emulate.c]: You can simply rewrite as following:
```c
exec_block_func_t _code = (exec_block_func_t) code;
if (unlikely(_code(rv, block->ir)))
```

File: src/riscv_private.h (hunk 38 above)
Comment (jserv) [reply src/riscv_private.h:156]: The following patch works for my Apple M1 machine:
```diff
--- a/src/riscv.c
+++ b/src/riscv.c
@@ -168,17 +168,6 @@ void rv_reset(riscv_t *rv, riscv_word_t pc)
     rv->halt = false;
 }
 
-#if defined(__APPLE__)
-static inline int isinff(float x)
-{
-    return __builtin_fabsf(x) == __builtin_inff();
-}
-static inline int isnanf(float x)
-{
-    return x != x;
-}
-#endif
-
 /* get current time in microsecnds and update csr_time register */
 static inline void update_time(riscv_t *rv)
 {
@@ -153,11 +153,15 @@ static inline uint32_t sign_extend_b(const uint32_t x)
 #include <math.h>
 #include "softfloat.h"
 
-#if defined(__APPLE__)
-inline int isinff(float x);
-inline int isnanf(float x);
+static inline int isinff(float x)
+{
+    return __builtin_fabsf(x) == __builtin_inff();
+}
+static inline int isnanf(float x)
+{
+    return x != x;
+}
 #endif
-#endif /* RV32_HAS(EXT_F) */
 
 #if RV32_HAS(Zicsr)
 /* CSRRW (Atomic Read/Write CSR) instruction atomically swaps values in the
```

Comment (qwe661234) [reply src/riscv_private.h:156]: It only works for the Apple architecture. In an x86 machine, the error occurs.
```
src/riscv_private.h:156:19: error: static declaration of 'isinff' follows non-static declaration
static inline int isinff(float x)
                  ^
src/riscv_private.h:160:19: error: static declaration of 'isnanf' follows non-static declaration
static inline int isnanf(float x)
                  ^
```

Comment (jserv) [reply src/riscv_private.h:156]: > It only works for the Apple architecture. In an x86 machine, the error occurs.

Consolidate the conditional build upon system configurations.

**Review by jserv — CHANGES_REQUESTED** [review]
1. Update CI pipeline to validate JIT-enabled execution.
2. Make comprehensive performance evaluations for various configurations. e.g., x86-64/linux, arm64/linux, arm64/macOS, etc.
3. Figure out performance bottleneck via profiling
4. Move forward to tracing JIT: https://en.wikipedia.org/wiki/Tracing_just-in-time_compilation

File: Makefile [hunk 40]
```diff
@@ -159,11 +184,11 @@ endif
 endif
 
 clean:
-	$(RM) $(BIN) $(OBJS) $(deps) $(CACHE_OUT)
+	$(RM) $(BIN) $(OBJS) $(deps) $(CACHE_OUT) src/jit_template.c
 distclean: clean
-	-$(RM) $(DOOM_DATA) $(QUAKE_DATA)
+	-$(RM) $(DOOM_DATA) $(QUAKE_DATA)
```
Comment (jserv) [inline Makefile]: Don't add trailing spaces.

File: src/cache.c [hunk 41]
```diff
@@ -8,6 +8,7 @@
 #include <stddef.h>
 #include <stdlib.h>
 #include <string.h>
+#include <sys/mman.h>
```
Comment (jserv) [inline src/cache.c:13]: Windows lacks of header `sys/mman.h`. Therefore, apply conditional build.

File: src/jit_template.c [hunk 42]
```diff
@@ -0,0 +1,511 @@
+RVOP(nop, { strcat(gencode, "/* no operation */\n"); })
```
Comment (jserv) [inline src/jit_template.c]: Don't check in `src/jit_template.c`.

File: Makefile [hunk 43]
```diff
@@ -78,6 +79,27 @@ gdbstub-test: $(BIN)
...
 
+ENABLE_JIT ?= 1
+$(call set-feature, JIT)
+ifeq ($(call has, JIT), 1)
+CFLAGS += -I./src/mir
+LDFLAGS += src/mir/libmir.a -lpthread
+OBJS_EXT += compile.o
+
+src/jit_template.c:
+	$(Q)python ./tools/gen-jit-template.py $(EXT_LIST) > src/jit_template.c
```
Comment (jserv) [inline Makefile]: Replace with
```shell
$(Q)python tools/gen-jit-template.py $(EXT_LIST) > $@
```

File: src/emulate.c [hunk 44]
```diff
@@ -34,6 +23,11 @@ extern struct target_ops gdbstub_ops;
 #include "riscv_private.h"
 #include "state.h"
 #include "utils.h"
+#if RV32_HAS(JIT)
+#include "cache.h"
+#include "compile.h"
+typedef bool (*exec_block_func_t)(riscv_t *rv, rv_insn_t *ir);
```
Comment (jserv) [inline src/emulate.c]: Don't typedef here.

File: src/emulate.c [hunk 45]
```diff
@@ -615,7 +504,26 @@ void rv_step(riscv_t *rv, int32_t cycles)
         }
         last_pc = rv->PC;
 
-        /* execute the block */
+#if RV32_HAS(JIT)
+        /* execute the block by JIT compiler */
+        exec_block_func_t code = NULL;
```
Comment (jserv) [inline src/emulate.c:512]: Move tyepdef `exec_block_func_t` here.

Comment (qwe661234) [reply src/emulate.c:512]: I typedef it to the top of function `rv_step`, because formatter format it to `typedef bool (*exec_block_func_t)(riscv_t * rv, rv_insn_t * ir);` if we tyepdef here.

File: Makefile [hunk 46]
```diff
@@ -13,6 +13,7 @@ CFLAGS += -D DEFAULT_STACK_ADDR=0xFFFFF000
 
 OBJS_EXT :=
 
+EXT_LIST :=
```
Comment (jserv) [inline Makefile]: To avoid potential confusion, let's rename it to `RVEXT_LIST` instead of `EXT_LIST`. The latter may mislead someone into thinking it represents a list of file extensions.

File: src/compile.c [hunk 47]
```diff
@@ -0,0 +1,411 @@
...
+    struct c2mir_options *options;
+    uint8_t debug_level;
+    uint8_t optimize_level;
+} riscv_jit_t;
+
+typedef struct {
+    char *name;
+    void *func;
+} func_obj_t;
+
```
Comment (jserv) [inline src/compile.c:37]: Remove this blank line.

File: src/compile.c [hunk 48]
```diff
@@ -0,0 +1,411 @@
...
+#define SET_SIZE 1 << SET_SIZE_BITS
+#define GOLDEN_RATIO_32 0x61C88647
+#define HASH(val) \
+    (((val) * (GOLDEN_RATIO_32)) >> (32 - SET_SIZE_BITS)) & ((SET_SIZE) -1)
+
+typedef struct {
+    uint32_t table[SET_SIZE][32];
+} set_t;
+
+static void set_reset(set_t *set)
```
Comment (jserv) [inline src/compile.c]: Decalare as a `static inline` function.

File: src/compile.c [hunk 49]
```diff
@@ -0,0 +1,411 @@
...
+{
+    memset(set, 0, sizeof(set_t));
+}
+
+static bool set_add(set_t *set, uint32_t key)
+{
+    uint32_t index = HASH(key);
+    uint8_t count = 0;
+    while (set->table[index][count] != 0) {
```
Comment (jserv) [inline src/compile.c]: Simply rename as
```c
while (set->table[index][count]) {
```

File: src/compile.c [hunk 50]
```diff
@@ -0,0 +1,411 @@
...
+
+    set->table[index][count] = key;
+    return true;
+}
+
+static bool set_has(set_t *set, uint32_t key)
+{
+    uint32_t index = HASH(key);
+    uint8_t count = 0;
+    while (set->table[index][count] != 0) {
```
Comment (jserv) [inline src/compile.c]: Ditto.

File: src/compile.c [hunk 51]
```diff
@@ -0,0 +1,411 @@
...
+#undef _
+        return true;
+    }
+    return false;
+}
+
+typedef void (*gen_func_t)(riscv_t *, const rv_insn_t *, char *, uint32_t);
+static gen_func_t dispatch_table[N_RV_INSN];
+static char funcbuf[128] = {0};
+#define APPEND_CODE(...)           \
```
Comment (jserv) [inline src/compile.c]: Rename it to `GEN`, which means generating code.

File: tools/gen-jit-template.py [hunk 52]
```diff
@@ -0,0 +1,209 @@
+import re
```
Comment (jserv) [inline tools/gen-jit-template.py:9]: Add a general description for this script.

File: src/compile.c [hunk 53]
```diff
@@ -0,0 +1,411 @@
...
+            APPEND_CODE("  rv->PC += ir->insn_len;\n");             \
+            APPEND_CODE("  ir = ir + 1;\n");                        \
+            NEXT_INSN(pc + ir->insn_len);                           \
+        }                                                           \
+    }
+
+#include "jit_template.c"
+
+RVOP(jal, {
+    strcat(gencode, "  pc = rv->PC;\n");
```
Comment (jserv) [inline src/compile.c]: Provide a macro to encapsulate, so that `gencode` won't appear here.

File: Makefile [hunk 54]
```diff
@@ -78,6 +79,27 @@ gdbstub-test: $(BIN)
...
 
+ENABLE_JIT ?= 0
+$(call set-feature, JIT)
+ifeq ($(call has, JIT), 1)
+CFLAGS += -I./src/mir
+LDFLAGS += src/mir/libmir.a -lpthread
+OBJS_EXT += compile.o
+
+src/jit_template.c:
+	$(Q)python ./tools/gen-jit-template.py $(EXT_LIST) > $@
```
Comment (jserv) [inline Makefile]: Rewrite as
```shell
$(Q)tools/gen-jit-template.py $(EXT_LIST) > $@
```

You shall specify a proper [shebang](https://en.wikipedia.org/wiki/Shebang_(Unix)) accordingly. i.e.,
```
#!/usr/bin/env python3
```

File: src/compile.c [hunk 55]
```diff
@@ -0,0 +1,411 @@
...
+        strcat(gencode, "  c = m->chunks[addr >> 16];\n");
+        APPEND_CODE(
+            "  rv->X[%u] = * (const uint32_t *) (c->data + (addr & "
+            "0xffff));\n",
+            fuse[i].rd);
+    }
+})
+#undef RVOP
+
+void trace_ebb(riscv_t *rv, char *gencode, rv_insn_t *ir, set_t *set)
```
Comment (jserv) [inline src/compile.c]: Declare it as `static`.

File: src/cache.c [hunk 56]
```diff
@@ -20,7 +25,13 @@
 /* THRESHOLD is set to identify hot spots. Once the frequency of use for a block
  * exceeds the THRESHOLD, the JIT compiler flow is triggered.
  */
-#define THRESHOLD 1000
+#define THRESHOLD 10000
```
Comment (jserv) [inline src/cache.c]: The change of `THRESHOLD` should come with an engineering consideration.

File: src/riscv.c [hunk 57]
```diff
@@ -116,11 +123,23 @@ bool rv_enables_to_output_exit_code(riscv_t *rv)
     return rv->output_exit_code;
 }
 
+#if RV32_HAS(JIT)
+void release_block(void *block)
```
Comment (jserv) [inline src/riscv.c]: Declare it as a `static` function.

File: src/cache.c (hunk 56 above)
Comment (qwe661234) [reply src/cache.c]: Perhaps we keep the value at 1000 for now and make modifications when we optimize the JIT.

File: src/compile.c [hunk 58]
```diff
@@ -0,0 +1,403 @@
...
+} riscv_jit_t;
+
+typedef struct {
+    char *name;
+    void *func;
+} func_obj_t;
+
+#define SET_SIZE_BITS 10
+#define SET_SIZE 1 << SET_SIZE_BITS
+#define GOLDEN_RATIO_32 0x61C88647
```
Comment (jserv) [inline src/compile.c]: Both `GOLDEN_RATIO_32` and `HASH` already exist in `src/cache.c`. We shall put them to somewhere such as `src/utils.h`.

**Comment by jserv** [comment]
It fails to build on macOS:
```shell
$ make  
Makefile:19: *** missing separator.  Stop.
```

System information:
```
$ make --version | head -1
GNU Make 3.81
```

File: src/compile.c [hunk 59]
```diff
@@ -0,0 +1,402 @@
...
+
+RVOP(sb, {MEMORY_FUNC(uint8_t, 0)})
+
+RVOP(sh, {MEMORY_FUNC(uint16_t, 0)})
+
+RVOP(sw, {MEMORY_FUNC(uint32_t, 0)})
+
+#if RV32_HAS(EXT_C)
+RVOP(clw, {
+    GEN("    addr = rv->X[%u] + %u;\n", ir->rs1, ir->imm);
```
Comment (jserv) [inline src/compile.c]: Be aware of the indention for C2MIR string.

File: src/riscv_private.h [hunk 60]
```diff
@@ -59,8 +61,12 @@ typedef struct block {
     uint32_t insn_capacity;    /**< maximum of instructions encompased */
     struct block *predict;     /**< block prediction */
     rv_insn_t *ir;             /**< IR as memory blocks */
+#if RV32_HAS(JIT)
+    bool hot;
```
Comment (jserv) [inline src/riscv_private.h]: Add corresponding comments here.

File: src/compile.c [hunk 61]
```diff
@@ -0,0 +1,402 @@
...
+        if (!insn_is_branch(ir->opcode)) {                          \
+            GEN("  rv->PC += ir->insn_len;\n");                     \
+            GEN("  ir = ir + 1;\n");                                \
+            NEXT_INSN(pc + ir->insn_len);                           \
+        }                                                           \
+    }
+
+#include "jit_template.c"
+
+RVOP(jal, {
```
Comment (jserv) [inline src/compile.c:118]: You have to explain why we have to provide the customized handlers for branch related instructions.

**Comment by qwe661234** [comment]
> It fails to build on macOS:
> 
> ```shell
> $ make  
> Makefile:19: *** missing separator.  Stop.
> ```
> 
> System information:
> 
> ```
> $ make --version | head -1
> GNU Make 3.81
> ```

Solved

**Comment by jserv** [comment]
Runtime failure on macOS/Arm64. Reproduce:
```shell
$ git clone https://github.com/sysprog21/rv32emu
$ cd rv32emu
$ gh pr checkout 135
$ make ENABLE_JIT=1
$ make check
```

Error messages:
```
Assertion failed: (cache->jitcode != MAP_FAILED), function cache_create, file cache.c, line 316.
Assertion failed: (cache->jitcode != MAP_FAILED), function cache_create, file cache.c, line 316.
Assertion failed: (cache->jitcode != MAP_FAILED), function cache_create, file cache.c, line 316.
Running hello.elf ... Failed.
```

**Comment by jserv** [comment]
This preliminary work was integrated into `wip/jit` branch for sake of synchronization with latest `master` branch.
@qwe661234, You shall work on `wip/jit` branch for all JIT specific issues and submit pull requests on top of it.
